sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala

```diff
@@ -0,0 +1,367 @@
+      require(metas.length == tables.length)
+
+      // copy the output data into correctly-sized buffers
+      metas.zipWithIndex.foreach { case (meta, i) =>
+        val oversizedBuffer = oversizedOutBuffers(i)
+        val compressedSize = meta.bufferMeta.size
+        val buffer = if (oversizedBuffer.getLength > compressedSize) {
+          oversizedBuffer.sliceWithCopy(0, compressedSize)
+        } else {
+          // use this buffer as-is, don't close it at the end of this method
+          oversizedOutBuffers(i) = null
+          oversizedBuffer
+        }
+        results += CompressedTable(compressedSize, meta, buffer)
+      }
+
+      // free all the inputs to this batch
+      tables.safeClose()
+      tables.clear()
+      tempBuffers.safeClose()
+      tempBuffers.clear()
+      oversizedOutBuffers.safeClose()
+      oversizedOutBuffers.clear()
+      batchMemUsed = 0
+    }
+  }
+
+  /** Return the amount of temporary space needed to compress this buffer */
+  protected def getTempSpaceNeeded(buffer: DeviceMemoryBuffer): Long
+
+  /** Return the amount of estimated output space needed to compress this buffer */
+  protected def getOutputSpaceNeeded(
+      dataBuffer: DeviceMemoryBuffer,
+      tempBuffer: DeviceMemoryBuffer): Long
+
+  /**
+   * Batch-compress contiguous tables
+   * @param outputBuffers output buffers allocated based on `getOutputSpaceNeeded` results
+   * @param tables contiguous tables to compress
+   * @param tempBuffers temporary buffers allocated based on `getTempSpaceNeeded` results.
+   *                    If the temporary space needed was zero then the corresponding buffer
+   *                    entry may be null.
+   * @return table metadata for the compressed tables. Table IDs should be set to 0.
+   */
+  protected def compress(
+      outputBuffers: Array[DeviceMemoryBuffer],
+      tables: Array[ContiguousTable],
+      tempBuffers: Array[DeviceMemoryBuffer]): Array[TableMeta]
+}
+
+/**
+ * Base class for batched decompressors
+ * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+ *                           which a batch should be compressed. A single table that requires
+ *                           temporary and output memory above this limit is allowed but will
+ *                           be compressed individually.
+ */
+abstract class BatchedBufferDecompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val inputBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val tempBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val outputBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val results = new ArrayBuffer[DeviceMemoryBuffer]
+
+  // temporary and output memory being used as part of the current batch
+  private[this] var batchMemUsed: Long = 0
+
+  /** The codec ID corresponding to this decompressor */
+  val codecId: Byte
```

Comment:
so here we refer to `codecId` instead of `codecType` so still a nit, but would be nice to pick one.